src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -379,38 +380,45 @@ def output_dict(self) -> Dict[str, Any]:
             }
         return self._output_dict
 
-    @cached_property
+    @property
     def plan_state(self) -> Dict[PlanState, Any]:
-        from snowflake.snowpark._internal.analyzer.select_statement import (
-            SelectStatement,
-        )
+        with self.session._plan_lock:
+            if self._plan_state is not None:
+                # return the cached plan state
```

Comment:
`@cached_property` is will allow multiple threads to concurrently evaluate the function thus creating race-conditions. This is something we want to avoid thus we do caching explicitly.